dbos-rules.test.ts

```diff
@@ -55,28 +55,244 @@ function makeExpectedDetCode(
   `;
 }
 
-function makeExpectedSuccessTest(code: string,
+function makeSqlInjectionCode(code: string): string {
+  return `
+    class UserDatabaseClient {}
+    class DBOSContext {}
+
+    class Knex {
+      raw(x: string) {
+
+      }
+    }
+
+    function Transaction(target: any, key: any, descriptor: any): any {
+      return descriptor;
+    }
+
+    export interface TransactionContext<T extends UserDatabaseClient> extends DBOSContext {
+      readonly client: T;
+    }
+
+    class Foo {
+      @Transaction()
+      injectionTime(ctxt: TransactionContext<Knex>, aParam: string) {
+        ${code}
+      }
+    }
+  `;
+}
+
+function errorIdsToObjectFormat(errorIds: string[]): { messageId: string }[] {
+  return errorIds.map((id) => { return { messageId: id }; });
+}
+
+function makeDeterminismSuccessTest(code: string,
   { codeAboveClass, enclosingFunctionParams } = { codeAboveClass: "", enclosingFunctionParams: "" }): SuccessTest {
 
-  return { code: makeExpectedDetCode(code, codeAboveClass, enclosingFunctionParams) };
+  return { code: makeDeterminismCode(code, codeAboveClass, enclosingFunctionParams) };
 }
 
-function makeExpectedFailureTest(code: string, expectedErrorIds: string[],
+function makeDeterminismFailureTest(code: string, expectedErrorIds: string[],
   { codeAboveClass, enclosingFunctionParams } = { codeAboveClass: "", enclosingFunctionParams: "" }): FailureTest {
 
-  const inObjectFormat = expectedErrorIds.map((id) => { return { messageId: id }; });
-  return { code: makeExpectedDetCode(code, codeAboveClass, enclosingFunctionParams), errors: inObjectFormat };
+  return {
+      code: makeDeterminismCode(code, codeAboveClass, enclosingFunctionParams),
+      errors: errorIdsToObjectFormat(expectedErrorIds)
+    };
+}
+
+function makeSqlInjectionSuccessTest(code: string): SuccessTest {
+  return { code: makeSqlInjectionCode(code) };
+}
+
+function makeSqlInjectionFailureTest(code: string, expectedErrorIds: string[]): FailureTest {
+  return {
+      code: makeSqlInjectionCode(code),
+      errors: errorIdsToObjectFormat(expectedErrorIds)
+    };
 }
 
 //////////
 
+// TODO: for the sake of testing the tests' soundness, check diagnostic warnings for these
 const testSet: TestSet = [
+  /* Note: the tests for SQL injection do not
+  involve any actual SQL code; they just test
+  for any non-LR-strings being passed to a raw SQL query callsite.
+  You can find more info on LR-strings in `dbos-rules.ts`. */
+
+  ["sql injection",
+    [
+      // Success test #1 (concatenation mania)
+      makeSqlInjectionSuccessTest(`
+        // Variables -> literals
+        const foo = "xyz", bar = "xyw";
+        ctxt.client.raw(foo);
+        ctxt.client.raw(bar);
+
+        // Variables -> variables -> ... -> literals
+
+        // Literal concatenation (TODO: make adding parentheses in not fail)
+        ctxt.client.raw("foo" + "bar" + "baz" + "bam");
+
+        // Literal + variable concatenation
```

Comment:
I would point out that this is OK because those variables have known values.

Reply:
Yes, it is - also, I fixed the parentheses thing, haven't pushed that yet